utils.py

```diff
@@ -0,0 +1,67 @@
+import contextlib
+import logging
+import re
+from pathlib import Path
+
+import pandas as pd
+from rich.logging import RichHandler
+
+
+def bulk_annotation_logger(log_level: str = "INFO"):
+    FORMAT = "%(message)s"
+
+    logging.basicConfig(level=log_level, format=FORMAT, datefmt="[%X]", handlers=[RichHandler()])
+
+    return logging.getLogger("rich")
+
+
+def output_dir() -> Path:
+    return Path(__file__).parent / "outputs"
+
+
+def dt_inplace(df: pd.DataFrame) -> pd.DataFrame:
+    """Automatically detect and convert (in place!) each dataframe column \
+    of datatype 'object' to a datetime just \
+    when ALL of its non-NaN values can be successfully parsed by pd.to_datetime().
+
+    Also returns a ref. to df for convenient use in an expression.
+
+    from :
+    https://towardsdatascience.com/auto-detect-and-set-the-date-datetime-datatypes-when-reading-csv-into-pandas-261746095361
+    """
+    from pandas.errors import ParserError
+
+    for c in df.columns[df.dtypes == "object"]:  # don't convert num
+        with contextlib.suppress(ParserError, ValueError, TypeError):
+            df[c] = pd.to_datetime(df[c])
+    return df
+
+
+def read_csv(*args, **kwargs) -> pd.DataFrame:
```

Comment:
Just a suggestion, but I'm wondering if it would be worth renaming this function to something a bit more descriptive so we don't forget its additional functionality when it's called elsewhere (and confuse it for `pd.read_csv`). Maybe `read_csv_with_dt`?

Reply:
agreed

Reply:
done